src/lib.rs

```diff
@@ -217,12 +212,37 @@ pub struct VirtualDestination {
 #[derive(PartialEq)]
 pub struct Device { object: Object }
 
+#[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+type AlignmentMarker = [u32; 0]; // ensures 4-byte alignment (on ARM)
+
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+type AlignmentMarker = [u8; 0]; // unaligned
+
 /// A [list of MIDI events](https://developer.apple.com/reference/coremidi/midipacketlist) being received from, or being sent to, one endpoint.
 ///
-#[derive(PartialEq)]
-pub struct PacketList(*const MIDIPacketList);
+#[repr(C)]
+pub struct PacketList {
+    // NOTE: This type must only exist in the form of immutable references
+    //       pointing to valid instances of MIDIPacketList.
+    //       This type must NOT implement `Copy`!
+    inner: PacketListInner,
+    _do_not_construct: AlignmentMarker
+}
+
+#[repr(packed)]
+struct PacketListInner {
+    num_packets: u32,
+    data: [MIDIPacket; 0]
```

Comment:
I am wondering if the differentiation between `PacketList` and `PacketListInner` is really necessary. Couldn't we just put the alignment marker after the `data` field and rename `PacketListInner` to `PacketList` ?

Reply:
Yes, `PacketListInner` is necessary, because of a lack of the `#[repr(packed = "N")]` feature in Rust (see https://github.com/rust-lang/rust/issues/33158). We want an alignment of 4 on ARM and 1 on other architectures, but `#[repr(packed)]` always gives an alignment of 1. So I'm defining `PacketListInner` with an alignment of 1, and using that within `PacketList` together with `AlignmentMarker`, so `PacketList` will choose the alignment as `max(alignof(PacketListInner), alignof(AlignmentMarker))`, which results in the correct alignment.